src/hera/workflows/task.py

```diff
@@ -42,6 +43,54 @@ class TaskResult(Enum):
     any_succeeded = "AnySucceeded"
     all_failed = "AllFailed"
 
+    def __or__(self, other: TaskResult) -> Iterable[TaskResult]:
+        """Create an "or" condition over multiple TaskResults."""
+        return _TaskResultGroup([self, other])
+
+
+class _TaskResultGroup(Iterable[TaskResult]):
+    """Private, implementation detail: an iterable of TaskResult with | support that maintains order and deduplicates."""
+
+    __slots__ = ("_results",)
+
+    def __init__(self, results: Iterable[TaskResult]):
+        self._results: List[TaskResult] = []
+        for r in results:
+            if r not in self._results:
+                self._results.append(r)
+
+    def __or__(self, other: Union[Iterable[TaskResult], TaskResult]):
+        if isinstance(other, TaskResult):
+            return _TaskResultGroup(self._results + [other])
+        return _TaskResultGroup(self._results + list(other))
+
+    def __iter__(self) -> Iterator[TaskResult]:
+        return iter(self._results)
+
+
+OnType = Optional[Union[TaskResult, Iterable[TaskResult]]]
+
+
+def _normalise_on(
+    on: OnType,
+    default: OnType = None,
+) -> Optional[List[TaskResult]]:
+    """Turn `on` into a list[TaskResult] or None.
+
+    Accepts:
+      - None -> return normalised default (which may also be None)
+      - TaskResult -> [TaskResult]
+      - Iterable[TaskResult] -> list(...)
+    """
+    if on is None:
+        if default is None:
+            return None
+
+        return _normalise_on(default)
```

Comment:
If I'm understanding the code correctly, this will always be normalized already.
```suggestion
        return default
```